File: FPG.Core/Algorithms/SymbolInsertion/NPickedSymbolInsertionAlgorithm.cs
using System;
using System.Text;

using FPG.Interfaces;

namespace FPG.Algorithms
{
    public class NPickedSymbolInsertionAlgorithm : ISymbolInsertionAlgorithm
    {
        public Int32 PickCount { get; set; } = 1;

        public String InsertSymbol(String input, String symbols)
        {
            StringBuilder sb = new StringBuilder(input);
            Int64 sum = 0;

            foreach (Char c in input)
            {
                sum += (Int64)c;
            }

            for (Int32 i = 0; i < this.PickCount; i++)
            {
                Int32 inputIndex = Convert.ToInt32((i + 1) * sum * 701 % input.Length);
                Int32 symbolIndex = Convert.ToInt32((i + 1) * sum * 997 % symbols.Length);

                // Prefer replacing digits (and not overwriting other symbols)
                Int32 tempInputIndex = inputIndex + 1;

                while (tempInputIndex != inputIndex && !Char.IsDigit(input[tempInputIndex]))
                {
                    tempInputIndex = (tempInputIndex + 1) % input.Length;
                }

                if (tempInputIndex == inputIndex)
                {
                    tempInputIndex = inputIndex + 1;

                    while (tempInputIndex != inputIndex && symbols.Contains(input[tempInputIndex].ToString()))
                    {
                        tempInputIndex = (tempInputIndex + 1) % input.Length;
                    }
                }

                if (tempInputIndex != inputIndex)
                {
                    inputIndex = tempInputIndex;
                }

                sb[inputIndex] = symbols[symbolIndex];
            }

            return sb.ToString();
        }
    }
}

File: lib/core/algorithms/symbolinsertion/NPickedSymbolInsertionAlgorithm.dart
import 'package:fpg_mobile/core/interfaces/ISymbolInsertionAlgorithm.dart';

class NPickedSymbolInsertionAlgorithm implements ISymbolInsertionAlgorithm {
  int pickCount = 1;

  @override
  String insertSymbol(String input, String? symbols) {
    StringBuffer sb = StringBuffer();
    // Dart does not support replace characters in place within StringBuffer
    // Use a map to do linear writing instead
    Map<int, int> indexMap = Map<int, int>();
    int sum = 0;

    input.runes.forEach((element) {
      sum += element;
    });

    for (int i = 0; i < this.pickCount; i++) {
      int inputIndex = ((i + 1) * sum * 701 % input.length).round();
      int symbolIndex = ((i + 1) * sum * 997 % symbols!.length).round();

      // Prefer replacing digits (and not overwriting other symbols)
      int alternativeInputIndex = inputIndex;

      while (alternativeInputIndex < input.length &&
          !isDigit(input, alternativeInputIndex)) {
        alternativeInputIndex++;
      }

      if (alternativeInputIndex == input.length) {
        alternativeInputIndex = inputIndex;

        while (alternativeInputIndex < input.length &&
            symbols.contains(input[alternativeInputIndex])) {
          alternativeInputIndex++;
        }
      }

      if (alternativeInputIndex < input.length &&
          alternativeInputIndex != inputIndex) {
        inputIndex = alternativeInputIndex;
      }

      indexMap[inputIndex] = symbolIndex;
    }

    for (int i = 0; i < input.length; i++) {
      sb.write(indexMap.containsKey(i) ? symbols![indexMap[i]!] : input[i]);
    }

    return sb.toString();
  }

  bool isDigit(String s, int idx) => (s.codeUnitAt(idx) ^ 0x30) <= 9;
}
